### PR TITLE
feat: Re enable restoration

### PR DESCRIPTION
**Pull request by adrien-coye**
### Abstract

Re-enable a state restoration in kDrive based on the UISceneDelegate refactor
Main PR to switch to UISceneDelegate https://github.com/Infomaniak/ios-kDrive/pull/1187

As discussed, this will focus on 
- [x] MainTabViewController
- [x] StoreViewController
- [x] FileDetailViewController
- [x] PreviewViewController
- [x] FileListViewController

This now uses the common way to load each screen. We navigate to it with a "router"
In the case of a state restoration, we do navigation calls on startup.

### CODE REVIEW

File: kDrive/UI/Controller/MainTabViewController.swift [hunk 1]
```diff
@@ -45,9 +45,21 @@ class MainTabViewController: UITabBarController, Restorable, PlusButtonObserver
         super.init(nibName: nil, bundle: nil)
         viewControllers = rootViewControllers
 
+        // Use selectedIndex argument if any
         if let selectedIndex {
             self.selectedIndex = selectedIndex
         }
+
+        // TODO: Abstract
```
Comment (adrien-coye) [inline kDrive/UI/Controller/MainTabViewController.swift:53]: TODO: This should be abstracted away from the init method, the correct index should be passed at init form reading the scene.

File: kDrive/SceneDelegate.swift [hunk 2]
```diff
@@ -87,10 +73,56 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate, AccountManagerDel
...
+
+        Log.sceneDelegate("restore from \(userActivity.activityType)")
+        Log.sceneDelegate("selectedIndex:\(userInfo["selectedIndex"])")
+    }
+
+    private func prepareWindowScene(_ windowScene: UIWindowScene) {
+        // Create a new UIWindow using the windowScene constructor which takes in a window scene.
+        let window = UIWindow(windowScene: windowScene)
+
+        // Create a view hierarchy programmatically
```
Comment (adrien-coye) [inline kDrive/SceneDelegate.swift:108]: TODO: clean

File: kDrive/AppRestorationService.swift [hunk 3]
```diff
@@ -59,13 +61,24 @@ public final class AppRestorationService: AppRestorationServiceable {
...
-        /* TODO: Rework app restoration before re-enabling
+        let encodedVersion = coder.decodeInteger(forKey: Self.appStateVersionKey)
+        let shouldRestoreApplicationState = Self.currentStateVersion == encodedVersion &&
+            !(UserDefaults.shared.legacyIsFirstLaunch || accountManager.accounts.isEmpty)
+        Log.appDelegate("shouldRestoreApplicationState:\(shouldRestoreApplicationState)")
+        return shouldRestoreApplicationState
+    }
+
+    public var shouldRestoreApplicationState: Bool {
+        return true
```
Comment (adrien-coye) [inline kDrive/AppRestorationService.swift:72]: TODO: change debug statement

**Review by adrien-coye — COMMENTED** [review]
Did a self assessment

File: kDrive/UI/Controller/Files/Preview/PreviewViewController.swift [hunk 4]
```diff
@@ -33,7 +33,27 @@ protocol PreviewContentCellDelegate: AnyObject {
     func openWith(from: UIView)
 }
 
-class PreviewViewController: UIViewController, PreviewContentCellDelegate {
+extension PreviewViewController: Encodable {
```
Comment (adrien-coye) [inline kDrive/UI/Controller/Files/Preview/PreviewViewController.swift:36]: TODO: Clean

File: kDrive/AppRouter.swift [hunk 5]
```diff
@@ -154,6 +183,139 @@ public struct AppRouter: AppNavigable {
...
+                guard let fromActivities = sceneUserInfo[SceneRestorationValues.fromActivities.rawValue] as? Bool else {
+                    fatalError("unable to load fromActivities")
+                }
+
+                let frozenFiles = database.fetchResults(ofType: File.self, filtering: { partial in
+                    partial.freezeIfNeeded()
+                })
+
+                // TODO: Use pred
+                let f = Array(frozenFiles).filter { fileIds.contains($0.id) }
```
Comment (adrien-coye) [inline kDrive/AppRouter.swift:295]: TODO clean

File: kDrive/AppRouter.swift [hunk 6]
```diff
@@ -44,18 +48,56 @@ public protocol RouterAppNavigable {
 
 /// Something that can present a File within the app
 public protocol RouterFileNavigable {
```
Comment (adrien-coye) [inline kDrive/AppRouter.swift:50]: TODO Document protocol

File: kDrive/SceneDelegate.swift [hunk 7]
```diff
@@ -259,7 +293,7 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate, AccountManagerDel
 // TODO: Refactor with router like pattern and split code away from this class
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDrive/SceneDelegate.swift:289]: not in the scope of this PR

File: kDriveCore/Data/Restoration/SceneRestorationKeys.swift [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    /// Array representing the stack of view controllers that should be restored
+    // TODO: Implement stack restoration
+    // case fileViewStack
+
+    /// The screen that should be restored on top of the MainViewController
+    case lastViewController
+}
+
+// TODO: Namespace by screen
+public enum SceneRestorationValues: String {
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDriveCore/Data/Restoration/SceneRestorationKeys.swift:49]: done

File: kDrive/AppRestorationService.swift [hunk 9]
```diff
@@ -59,13 +62,27 @@ public final class AppRestorationService: AppRestorationServiceable {
...
+             !(UserDefaults.shared.legacyIsFirstLaunch || accountManager.accounts.isEmpty)
+         Log.appDelegate("shouldRestoreApplicationState:\(shouldRestoreApplicationState)")
+         return shouldRestoreApplicationState */
+    }
+
+    public var shouldRestoreApplicationState: Bool {
+        print("•• shouldRestoreApplicationState")
+        return true
+
+        // TODO: Use user defaults
```
[the remark this answers is not in the document]
Comment (adrien-coye) [reply kDrive/AppRestorationService.swift:78]: Will do in another PR

File: kDrive/SceneDelegate.swift [hunk 10]
```diff
@@ -87,10 +73,49 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate, AccountManagerDel
...
+
+    private func prepareWindowScene(_ windowScene: UIWindowScene) {
+        // Create a new UIWindow using the windowScene constructor which takes in a window scene.
+        let newWindow = UIWindow(windowScene: windowScene)
+
+        // Set the window and call makeKeyAndVisible()
+        window = newWindow
+        newWindow.makeKeyAndVisible()
+
+        // Update tint
```
Comment (PhilippeWeidmann) [inline kDrive/SceneDelegate.swift:112]:
```suggestion
```

File: kDrive/SceneDelegate.swift [hunk 11]
```diff
@@ -55,26 +56,11 @@ final class SceneDelegate: UIResponder, UIWindowSceneDelegate, AccountManagerDel
          The recommended approach is for the SceneDelegate to retain the scene's window.
      */
     func scene(_ scene: UIScene, willConnectTo session: UISceneSession, options connectionOptions: UIScene.ConnectionOptions) {
-        print(" scene session options")
+        Log.sceneDelegate("scene session options")
         /// 1. Capture the scene
```
Comment (PhilippeWeidmann) [inline kDrive/SceneDelegate.swift:60]:
```suggestion
```

Comment (adrien-coye) [reply kDrive/SceneDelegate.swift]: Sure I'll stick to our standard which is "only protocols, or bugs and oddities"
I code this way, I'll make sure to match our rule before asking for review.

That said, now that we split the code in more SRP fashion, I do feel like we can get away with less comments.

Comment (adrien-coye) [reply kDrive/SceneDelegate.swift]: > it was part of a series of comments 1. 2. 3. (that I probably didn't notice in an earlier review) but now it's just sitting here.

For the record, this was part of my SceneDelegate PoC (the main PR) and was ought to be removed.
Is never was "upstreamed" to releaseAPIV3

File: kDrive/AppRouter.swift (hunk 6 above)
Comment (adrien-coye) [reply kDrive/AppRouter.swift:50]: done

**Review by PhilippeWeidmann — APPROVED** [review]

**Pull request merged**